### DISCUSSION TITLE
The new OptionsValidator (for validating configuration values) is only run time (when first accessed) and not during startup?

### DISCUSSION THREAD

**Discussion by PureKrome** [dotnet/runtime#96729] [Q&A] [answered]
With .NET 8, there is now the new way to read in configuration source data (data from config files or env vars, keyvault, etc) => using a new [IOptionsValidator class](https://learn.microsoft.com/en-us/dotnet/core/extensions/options-validation-generator). This is great because the options validation source generator can produce optimized validation code at compile time.

But the actual _values_ are checked at _run time_ when the IOptions instance is first 'required' .. and not during the OnStartup. which means, if a piece of code which requires some IOptions isn't 'called' for a while, then we don't know the configuration data is "bad".

Are there any plans to extend this to optionally wire this up to OnStartup? I noticed that there is an [IStartUpValidator interface already](https://github.com/dotnet/runtime/blob/cc5f1df48e5e39a6fa5ab94ca4664aac3cc36898/src/libraries/Microsoft.Extensions.Options/src/IStartupValidator.cs).

Or maybe this exists and the documentation is lacking/missing mentioning this?

**Comment by gfoidl** [comment, marked as answer]
Did you see [AddOptionsWithValidateOnStart](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.optionsservicecollectionextensions.addoptionswithvalidateonstart?view=dotnet-plat-ext-8.0)?

Relevant issue for this: https://github.com/dotnet/runtime/issues/89263

**Reply by PureKrome** [reply]
![giphy 1](https://github.com/dotnet/runtime/assets/899878/ecc06252-6794-46d0-aaa9-a80a54e9504c)

WAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA!!!!

omg yes! secret method that I've never see in any documentation _but it exists!_

Thank you so much @gfoidl - maybe this should be highlighted in some of the "[getting started with options](https://learn.microsoft.com/en-us/dotnet/core/extensions/options)" documentation? Where would I ask to see if a mention of the `AddOptionsWithValidateOnStart` should be mentioned in that doc?

again - thanks heaps Günther!

**Reply by gfoidl** [reply]
> Where would I ask to see if a mention of the AddOptionsWithValidateOnStart should be mentioned in that doc?

At the bottom of the linked docs article is a link "Open a documentation issue", there you can file an issue (and as that's a valuable addition I'm pretty sure it's picked up).

**Reply by PureKrome** [reply]
thanks heaps @gfoidl - will do. 🍰 Ta mate!

**Reply by PureKrome** [reply]
[Doc issue opened up here](https://github.com/dotnet/runtime/discussions/96729#discussioncomment-8088337).

**Reply by gfoidl** [reply]
The link here is the comment itself. Here's the docs issue: https://github.com/dotnet/docs/issues/39089
Thanks for filing it.

**Reply by PureKrome** [reply]
oh lol. soz. i copied the wrong link 😊 - thanks for pasting the right one! cheers!
